include/evmc.h

```diff
@@ -140,48 +140,97 @@ typedef void (*evmc_get_block_hash_fn)(struct evmc_uint256be* result,
+     * The execution has attempted to put more items on the EVM stack
+     * than the specified limit.
+     */
+    EVMC_STACK_OVERFLOW = 4,
+
+
+    /** Execution of an opcode has required more items on the EVM stack. */
+    EVMC_STACK_UNDERFLOW = 5,
+
+    /** Execution has violated the jump destination restrictions. */
+    EVMC_BAD_JUMP_DESTINATION = 6,
+
+    /** An undefined instruction has been encountered. */
+    EVMC_UNDEFINED_INSTRUCTION = 7,
+
+    /**
+     * The designated INVALID instruction has been hit during execution.
+     *
+     * The EIP-141 (https://github.com/ethereum/EIPs/blob/master/EIPS/eip-141.md)
+     * defines the instruction 0xfe as INVALID instruction to indicate execution
+     * abortion coming from high-level languages. This status code is reported
+     * in case this INVALID instruction has been encountered.
+     */
+    EVMC_INVALID_INSTRUCTION = 8,
+
+    /** Tried to execute an operation which is restricted in static mode. */
+    EVMC_STATIC_MODE_VIOLATION = 9,
+
+    /**
+     * Tried to read outside memory bounds.
+     *
+     * An example is RETURNDATACOPY reading past the available buffer.
+     */
     EVMC_INVALID_MEMORY_ACCESS = 10,
 
-    /// Exceptions produced by precompiles/system contracts
-    ///
-    /// An example: elliptic curve functions handed invalid EC points
+    /**
+     * A call to a precompiled or system contract has ended with a failure.
+     *
+     * An example: elliptic curve functions handed invalid EC points.
+     */
     EVMC_PRECOMPILE_FAILURE = 11,
 
-    /// The EVM rejected the execution of the given code or message.
-    ///
-    /// This error SHOULD be used to signal that the EVM is not able to or
-    /// willing to execute the given code type or message.
-    /// If an EVM returns the ::EVMC_REJECTED status code,
-    /// the Client MAY try to execute it in other EVM implementation.
-    /// For example, the Client tries running a code in the EVM 1.5. If the
-    /// code is not supported there, the execution falls back to the EVM 1.0.
-    EVMC_REJECTED = -1,
-
-    /// EVM implementation internal error.
-    ///
-    /// @todo We should rethink reporting internal errors. One of the options
-    ///       it to allow using any negative value to represent internal errors.
-    EVMC_INTERNAL_ERROR = -2,
+
+    /** EVM implementation generic internal error. */
+    EVMC_INTERNAL_ERROR = -1,
+
+    /**
+     * The execution of the given code and/or message has been rejected
+     * by the EVM implementation.
+     *
+     * This error SHOULD be used to signal that the EVM is not able to or
+     * willing to execute the given code type or message.
+     * If an EVM returns the ::EVMC_REJECTED status code,
+     * the Client MAY try to execute it in other EVM implementation.
+     * For example, the Client tries running a code in the EVM 1.5. If the
+     * code is not supported there, the execution falls back to the EVM 1.0.
+     */
+    EVMC_REJECTED = -2,
```

Comment:
What is the reason for swapping these two?

Reply:
The generic first, the specific after.

Reply:
I a bit torn on this, since `REJECTED` is a special one and I am not sure it is an internal error. It is certainly recoverable.

Reply:
It's internal in sense it should not be passed back to the caller.
Yes, it's recoverable if the client supports it and has a backup VM. I will extend the description of `evmc_status_code` to mention this as suggested.